src/mc.rs

```diff
@@ -1004,78 +1006,72 @@ unsafe fn emu_edge_rust<BD: BitDepth>(
     ih: intptr_t,
     x: intptr_t,
     y: intptr_t,
-    mut dst: *mut BD::Pixel,
-    dst_stride: ptrdiff_t,
+    dst: *mut [BD::Pixel; EMU_EDGE_LEN],
+    dst_stride: usize,
     mut r#ref: *const BD::Pixel,
     ref_stride: ptrdiff_t,
 ) {
-    r#ref = r#ref.offset(
-        iclip(y as c_int, 0 as c_int, ih as c_int - 1) as isize * BD::pxstride(ref_stride)
-            + iclip(x as c_int, 0 as c_int, iw as c_int - 1) as isize,
-    );
-    let left_ext = iclip(-x as c_int, 0 as c_int, bw as c_int - 1);
-    let right_ext = iclip((x + bw - iw) as c_int, 0 as c_int, bw as c_int - 1);
-    if !(((left_ext + right_ext) as isize) < bw) {
-        unreachable!();
-    }
-    let top_ext = iclip(-y as c_int, 0 as c_int, bh as c_int - 1);
-    let bottom_ext = iclip((y + bh - ih) as c_int, 0 as c_int, bh as c_int - 1);
-    if !(((top_ext + bottom_ext) as isize) < bh) {
-        unreachable!();
-    }
-    let mut blk: *mut BD::Pixel =
-        dst.offset((top_ext as isize * BD::pxstride(dst_stride)) as isize);
-    let center_w = (bw - left_ext as isize - right_ext as isize) as c_int;
-    let center_h = (bh - top_ext as isize - bottom_ext as isize) as c_int;
-    let mut y_0 = 0;
-    while y_0 < center_h {
+    let dst = &mut *dst;
+    let dst_stride = BD::pxstride(dst_stride);
+    let ref_stride = BD::pxstride(ref_stride);
+
+    // find offset in reference of visible block to copy
+    r#ref = r#ref.offset(clip(y, 0, ih - 1) * ref_stride + clip(x, 0, iw - 1));
+
+    // number of pixels to extend (left, right, top, bottom)
+    let left_ext = clip(-x, 0, bw - 1) as usize;
+    let right_ext = clip(x + bw - iw, 0, bw - 1) as usize;
+    assert!(((left_ext + right_ext) as isize) < bw);
+    let top_ext = clip(-y, 0, bh - 1) as usize;
+    let bottom_ext = clip(y + bh - ih, 0, bh - 1) as usize;
+    assert!(((top_ext + bottom_ext) as isize) < bh);
+
+    let bw = bw as usize;
+    let bh = bh as usize;
+
+    // copy visible portion first
+    let mut blk = top_ext * dst_stride;
+    let center_w = bw - left_ext - right_ext;
+    let center_h = bh - top_ext - bottom_ext;
+    for _ in 0..center_h {
         BD::pixel_copy(
-            std::slice::from_raw_parts_mut(blk.offset(left_ext as isize), center_w as usize),
-            std::slice::from_raw_parts(r#ref, center_w as usize),
-            center_w as usize,
+            &mut dst[blk + left_ext..][..center_w],
+            std::slice::from_raw_parts(r#ref, center_w),
+            center_w,
         );
+        // extend left edge for this line
         if left_ext != 0 {
-            BD::pixel_set(
-                std::slice::from_raw_parts_mut(blk, left_ext as usize),
-                *blk.offset(left_ext as isize),
-                left_ext as usize,
-            );
+            let val = dst[blk + left_ext];
+            dst[blk..][..left_ext].fill(val);
```

Comment:
```suggestion
            BD::pixel_set(&dst[blk..], val, left_ext);
```

I know these do the same thing, but maybe we should stick with the naming C uses?  That said, this way is indeed easier to read, so your call.

Reply:
Eh yeah I can go back to using `pixel_set`, it doesn't matter much either way.